src/main/java/com/google/devtools/build/lib/remote/common/RemotePathResolver.java

```diff
@@ -0,0 +1,194 @@
+// Copyright 2020 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.google.devtools.build.lib.remote.common;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.devtools.build.lib.actions.ActionInput;
+import com.google.devtools.build.lib.actions.ActionInputHelper;
+import com.google.devtools.build.lib.exec.SpawnRunner.SpawnExecutionContext;
+import com.google.devtools.build.lib.vfs.Path;
+import com.google.devtools.build.lib.vfs.PathFragment;
+import java.io.IOException;
+import java.util.SortedMap;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link RemotePathResolver} is used to resolve input/output paths for remote execution from
+ * Bazel's internal path, or vice versa.
+ */
+public interface RemotePathResolver {
+
+  /**
+   * @return the {@code workingDirectory} for a remote action.
+   */
+  @Nullable
+  String getWorkingDirectory();
+
+  /**
+   * @return a {@link SortedMap} which maps from input paths for remote action to {@link
+   * ActionInput}.
+   */
+  SortedMap<PathFragment, ActionInput> getInputMapping(SpawnExecutionContext context)
+      throws IOException;
+
+  /**
+   * Resolves the output path relative to input root for the given {@link Path}.
+   */
+  String resolveOutputPath(Path path);
+
+  /**
+   * Resolves the output path relative to input root for the given {@link PathFragment}.
+   *
+   * @param execPath a path fragment relative to {@code execRoot}.
+   */
+  String resolveOutputPath(PathFragment execPath);
+
+  /**
+   * Resolves the output path relative to input root for the {@link ActionInput}.
+   */
+  default String resolveOutputPath(ActionInput actionInput) {
+    return resolveOutputPath(actionInput.getExecPath());
+  }
+
+  /**
+   * Resolves the local {@link Path} of an output file.
+   *
+   * @param outputPath the return value of {@link #resolveOutputPath(PathFragment)}.
+   */
+  Path resolveLocalPath(String outputPath);
```

Comment:
opt: maybe it's better to call these methods `localPathToOutputPath` or `outputPathToLocalPath` because the verb "resolve" is quite vague.

Reply:
Done.